File: 0057-insert-interval/0057-insert-interval.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        List<int[]> list = new ArrayList<>();
        int n = intervals.length, i = 0;
        //Loop until the end element of interval is less than the first element of newInterval
        while(i<n && intervals[i][1] < newInterval[0]){
            //keep adding the intervals to list
            list.add(intervals[i]);
            i++;
        }
        //now we can merge the newInterval to intervals
        //check until when is the end of newInterval greater than the start of interval item, update newInterval based on the min and max value
        while(i<n && newInterval[1]>=intervals[i][0]){
            newInterval[0]=Math.min(intervals[i][0], newInterval[0]);
            newInterval[1]=Math.max(intervals[i][1], newInterval[1]);
            i++;
        }
        //add it to list
        list.add(newInterval);
        //add remaining intervals
        while(i<n){
            list.add(intervals[i]);
            i++;
        }
        return list.toArray(new int[list.size()][2]);
    }
}

File: 0057-insert-interval/0057-insert-interval.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        res = []

        for i in range(len(intervals)):
            #if new interval is strictly before current interval append to list and return remaining items
            if newInterval[1] < intervals[i][0]:
                res.append(newInterval)
                return res + intervals[i:]
            #if new interval is strictly after current interval append current interval
            elif newInterval[0] > intervals[i][1]:
                res.append(intervals[i])
            #if the intervals are overlapping merge them and update the newInterval
            else:
                newInterval = [min(newInterval[0], intervals[i][0]), max(newInterval[1], intervals[i][1])]

        res.append(newInterval)
        return res